### PR TITLE
EN-72: Contract status based on Due date

### PR DESCRIPTION
**Pull request by FacuOrellana**
# Description :pencil:

The contract status now is based on due date

## Link to ticket :link:

https://entropyteam.atlassian.net/browse/EN-72

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? :microscope:

Manual testing.

the following cases were tested:

the following cases were tested:

1. New active contract: when a new valid contract is created it is active

2. New inactive contract: when a new contract has end date before now it is inactive

3. New contract with end date null is active: when a new contract has null end date it is active

4. New contract with start date after now and end date after start date is inactive

5. Inactive last active contract when a new active contract is created: when a new active contract is created the last active contract is set inactive

6. Last active contract stays active when inactive contract is created

7. Last active contract set inactive when new contract with null end date is created

### CODE REVIEW

File: src/main/java/com/entropyteam/entropay/employees/repositories/ContractRepository.java [hunk 1]
```diff
@@ -9,5 +9,5 @@
 public interface ContractRepository extends BaseRepository<Contract, UUID> {
 
     List<Contract> findAllByDeletedIsFalse();
-    Optional<Contract> findContractByEmployeeIdAndActiveIsTrue(UUID employeeId);
+    Optional<Contract> findContractByEmployeeIdAndActiveIsTrueAndDeletedIsFalse(UUID employeeId);
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/repositories/ContractRepository.java:12]: Be careful that there is no other method using this

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 2]
```diff
@@ -60,16 +63,10 @@ public ContractService(ContractRepository contractRepository, CompanyRepository
...
-                    existent.setActive(false);
-                    existent.setModifiedAt(LocalDateTime.now());
-                    existent.setEndDate(LocalDate.now());
-                    contractRepository.saveAndFlush(existent);
-                });
-        Contract entityToCreate = toEntity(contractDto.withActive(true));
+        Contract entityToCreate = toEntity(contractDto);
         Contract savedEntity = getRepository().save(entityToCreate);
-        paymentSettlementService.create(contractDto.paymentSettlement(), savedEntity);
+        paymentSettlementService.create(savedEntity.getPaymentsSettlement(), savedEntity);
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:68]: is savedEntity populated with the payment settlements?

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 3]
```diff
@@ -60,16 +63,10 @@ public ContractService(ContractRepository contractRepository, CompanyRepository
...
-        contractRepository.findContractByEmployeeIdAndActiveIsTrue(contractDto.employeeId())
-                .ifPresent(existent -> {
-                    existent.setActive(false);
-                    existent.setModifiedAt(LocalDateTime.now());
-                    existent.setEndDate(LocalDate.now());
-                    contractRepository.saveAndFlush(existent);
-                });
-        Contract entityToCreate = toEntity(contractDto.withActive(true));
+        Contract entityToCreate = toEntity(contractDto);
         Contract savedEntity = getRepository().save(entityToCreate);
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:67]: i would save the contract at the end of the method

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 4]
```diff
@@ -136,7 +134,23 @@ protected Contract toEntity(ContractDto entity) {
         contract.setEmployee(employee);
         contract.setRole(role);
         contract.setSeniority(seniority);
+        if(entity.paymentSettlement() == null) contract.setPaymentsSettlement(new HashSet<>());
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:137]: Please change to:
```
if(entity.paymentSettlement() == null){
     contract.setPaymentsSettlement(new HashSet<>());
 }
 ```

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 5]
```diff
@@ -136,7 +134,23 @@ protected Contract toEntity(ContractDto entity) {
...
 
         return contract;
     }
+
+    protected void checkActiveContract(Contract contractToCheck) {
+        Optional<Contract> activeContract = contractRepository.findContractByEmployeeIdAndActiveIsTrueAndDeletedIsFalse(contractToCheck.getEmployee().getId());
+        if ((contractToCheck.getEndDate() == null || contractToCheck.getEndDate().isAfter(LocalDate.now())) && (contractToCheck.getStartDate().isBefore(LocalDate.now())) || contractToCheck.getStartDate().isEqual(LocalDate.now()) ) {
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:144]: `(contractToCheck.getStartDate().isBefore(LocalDate.now())) || contractToCheck.getStartDate().isEqual(LocalDate.now())` = `!contractToCheck.getStartDate().isAfter(LocalDate.now()`

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 6]
```diff
@@ -136,7 +134,23 @@ protected Contract toEntity(ContractDto entity) {
...
+            contractToCheck.setActive(true);
+            activeContract.ifPresent(contract -> {
+                        contract.setActive(false);
+                        if (contract.getEndDate() == null) contract.setEndDate(LocalDate.now());
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:148]: please format this if with {} as above

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 7]
```diff
@@ -136,7 +134,23 @@ protected Contract toEntity(ContractDto entity) {
         contract.setEmployee(employee);
         contract.setRole(role);
         contract.setSeniority(seniority);
+        if(entity.paymentSettlement() == null) contract.setPaymentsSettlement(new HashSet<>());
 
         return contract;
     }
+
+    protected void checkActiveContract(Contract contractToCheck) {
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:142]: this method should return the contract then save it in the caller method

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java (hunk 5 above)
Comment (FacuOrellana) [reply src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:144]: does this conditions checks if end date is after today?

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 8]
```diff
@@ -136,7 +132,26 @@ protected Contract toEntity(ContractDto entity) {
...
         return contract;
     }
+
+    protected Contract checkActiveContract(Contract contractToCheck) {
+        Optional<Contract> activeContract = contractRepository.findContractByEmployeeIdAndActiveIsTrueAndDeletedIsFalse(contractToCheck.getEmployee().getId());
+        if ((contractToCheck.getEndDate() == null || contractToCheck.getEndDate().isAfter(LocalDate.now())) && (contractToCheck.getStartDate().isBefore(LocalDate.now())) || contractToCheck.getStartDate().isEqual(LocalDate.now()) == !contractToCheck.getStartDate().isAfter(LocalDate.now())) {
+            contractToCheck.setActive(true);
+            activeContract.ifPresent(contract -> {
+                            contract.setActive(false);
+                            if (contract.getEndDate() == null) contract.setEndDate(LocalDate.now());
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:149]: fix format

File: src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java [hunk 9]
```diff
@@ -3,14 +3,11 @@
 import com.entropyteam.entropay.common.BaseRepository;
 import com.entropyteam.entropay.common.BaseService;
 import com.entropyteam.entropay.common.ReactAdminMapper;
+import com.entropyteam.entropay.employees.models.*;
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java:6]: fix import

File: src/test/java/com/entropyteam/entropay/employees/services/ContractServiceTest.java [hunk 10]
```diff
@@ -156,12 +156,13 @@ public void testCreateWhenContractRepositoryThrowsException() {
         ContractDto requestedContract = new ContractDto(existentContract);
 
         // when
-        when(contractRepository.findContractByEmployeeIdAndActiveIsTrue(any())).thenThrow(
-                new RuntimeException("Test exception thrown!!"));
+        // when(contractRepository.findContractByEmployeeIdAndActiveIsTrueAndDeletedIsFalse(any())).thenThrow(
```
Comment (agustinb4) [inline src/test/java/com/entropyteam/entropay/employees/services/ContractServiceTest.java:159]: please delete comments

File: src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java [hunk 11]
```diff
@@ -66,6 +66,11 @@ protected Employee toEntity(EmployeeDto entity) {
 
         employee.setRoles(roles);
         employee.setTechnologies(technologies);
+        if (entity.paymentInformation() == null) {
+            employee.setPaymentsInformation(new HashSet<>());
+        } else {
+            employee.setPaymentsInformation(new HashSet<>(entity.paymentInformation().stream().map(PaymentInformation::new).toList()));
+        }
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java:73]: what's about:
`employee.setPaymentsInformation(entity.paymentInformation() == null ?  Collections.EMPTY_SET : entity.paymentInformation().stream().map(PaymentInformation::new).toSet())`

File: src/main/java/com/entropyteam/entropay/employees/services/ContractService.java [hunk 12]
```diff
@@ -136,7 +132,28 @@ protected Contract toEntity(ContractDto entity) {
         contract.setEmployee(employee);
         contract.setRole(role);
         contract.setSeniority(seniority);
-
+        if (entity.paymentSettlement() == null) {
+            contract.setPaymentsSettlement(new HashSet<>());
+        } else {
+            contract.setPaymentsSettlement(new HashSet<>(entity.paymentSettlement().stream().map(PaymentSettlement::new).toList()));
+        }
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/ContractService.java:139]: what's about:
`employee.setPaymentsSettlement(entity.paymentSettlement() == null ?  Collections.EMPTY_SET : entity.paymentSettlement().stream().map(PaymentSettlement::new).toSet())`

File: src/test/java/com/entropyteam/entropay/employees/services/ContractServiceTest.java [hunk 13]
```diff
@@ -343,8 +344,155 @@ public void testModifyStatusWhenContractRepositoryThrowsException() {
 
         // then
         RuntimeException thrown = assertThrows(RuntimeException.class, () ->
-                        underTest.modifyStatus(existentContract.getId(), setActive),
+                        contractService.modifyStatus(existentContract.getId(), setActive),
                 "RuntimeException was expected");
         verifyNoMoreInteractions(contractRepository);
     }
+
```
Comment (agustinb4) [inline src/test/java/com/entropyteam/entropay/employees/services/ContractServiceTest.java:351]: Nice!

**Review by agustinb4 — APPROVED** [review]
Looks good, good job with those tests

**Pull request merged**